File: python/2015/day07b/day07b.py
#!/usr/bin/env python3

import re
import sys

def usage():
    progname = sys.argv[0]
    print(f'usage: {progname} <input file>')
    sys.exit(1)

def process(filename):
    operations = {}
    pattern = re.compile(r'(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)')
    p1 = re.compile(r'^(\d+|\w+) -> (\w+)')
    p2 = re.compile(r'NOT (\d+|\w+) -> (\w+)')
    p3 = re.compile(r'(\d+|\w+) (AND|OR) (\d+|\w+) -> (\w+)')
    p4 = re.compile(r'(\d+|\w+) (LSHIFT|RSHIFT) (\d+) -> (\w+)')
    with open(filename, 'r') as infile:
        for line in infile:
            matches = p1.match(line)
            if matches != None:
                src = matches[1]
                dest = matches[2]
                operations[dest] = ('ASSIGN', src)
            matches = p2.match(line)
            if matches != None:
                src = matches[1]
                dest = matches[2]
                operations[dest] = ('NOT', src)
            matches = p3.match(line)
            if matches != None:
                src1 = matches[1]
                op = matches[2]
                src2 = matches[3]
                dest = matches[4]
                operations[dest] = (op, src1, src2)
            matches = p4.match(line)
            if matches != None:
                src = matches[1]
                op = matches[2]
                amt = int(matches[3])
                dest = matches[4]
                operations[dest] = (op, src, amt)
    a = evaluate(operations, {}, "a")
    operations["b"] = ("ASSIGN", str(a))
    return evaluate(operations, {}, "a")

def evaluate(operations, cache, expr):
    try:
        return int(expr)
    except ValueError:
        pass
    if ( expr in cache ):
        return cache[expr]
    else:
        op = operations[expr]
        operator = op[0]
        if operator == 'ASSIGN':
            src = op[1]
            a = evaluate(operations, cache, src)
            r = a
        elif operator == 'NOT':
            src = op[1]
            a = evaluate(operations, cache, src)
            r = ~a
        elif operator == 'AND':
            src1 = op[1]
            src2 = op[2]
            a = evaluate(operations, cache, src1)
            b = evaluate(operations, cache, src2)
            r = a & b
        elif operator == 'OR':
            src1 = op[1]
            src2 = op[2]
            a = evaluate(operations, cache, src1)
            b = evaluate(operations, cache, src2)
            r = a | b
        elif operator == 'LSHIFT':
            src = op[1]
            amt = op[2]
            a = evaluate(operations, cache, src)
            r = a << amt
        elif operator == 'RSHIFT':
            src = op[1]
            amt = op[2]
            a = evaluate(operations, cache, src)
            r = a >> amt
        else:
            raise RuntimeError
        masked = r & 0xffff
        cache[expr] = masked
        return masked

def main():
    if len(sys.argv) < 2:
        usage()
    filename = sys.argv[1]
    result = process(filename)
    print(f'result = {result}')

if __name__ == '__main__':
    main()

File: perl/2015/day07b/day07b.pl
#!/usr/bin/env perl

use strict;
use warnings;

sub usage {
    my $progname = $0;
    print "usage: $progname <input file>\n";
    exit 1;
}

sub process {
    my ($filename) = @_;
    my %operations = ();
    my %cache = ();

    open ( my $infile, '<', $filename ) or die "cannot open file: $filename: $!";
    while ( my $line = <$infile> ) {
        if ( $line =~ /^(\d+|\w+) -> (\w+)$/ ) {
            my ($src, $dest) = ($1, $2);
            $operations{$dest} = ["ASSIGN", $src];
        } elsif ( $line =~ /NOT (\d+|\w+) -> (\w+)/ ) {
            my ($src, $dest) = ($1, $2);
            $operations{$dest} = ["NOT", $src];
        } elsif ( $line =~ /(\d+|\w+) (AND|OR) (\d+|\w+) -> (\w+)/ ) {
            my ($src1, $op, $src2, $dest) = ($1, $2, $3, $4);
            $operations{$dest} = [$op, $src1, $src2];
        } elsif ( $line =~ /(\d+|\w+) (LSHIFT|RSHIFT) (\d+) -> (\w+)/ ) {
            my ($src, $op, $amt, $dest) = ($1, $2, $3, $4);
            $operations{$dest} = [$op, $src, $amt];
        } else {
            next;
        }
    }
    close($infile);

    my $a = evaluate(\%operations, \%cache, "a");
    $operations{"b"} = ["ASSIGN", $a];
    %cache = ();
    return evaluate(\%operations, \%cache, "a");
}

sub evaluate {
    my ($ops, $cache, $expr) = @_;

    if ( $expr =~ /\d+/ ) {
        return $expr;
    } elsif ( exists $cache->{$expr} ) {
        return $cache->{$expr};
    } else {
        my @op = @{ $ops->{$expr} };
        my $r = 0;
        my $operator = $op[0];
        if ( $operator eq "ASSIGN" ) {
            my $src = $op[1];
            my $a = evaluate($ops, $cache, $src);
            $r = $a;
        } elsif ( $operator eq "NOT" ) {
            my $src = $op[1];
            my $a = evaluate($ops, $cache, $src);
            $r = ~$a;
        } elsif ( $operator eq "AND" ) {
            my $src1 = $op[1];
            my $src2 = $op[2];
            my $a = evaluate($ops, $cache, $src1);
            my $b = evaluate($ops, $cache, $src2);
            $r = $a & $b;
        } elsif ( $operator eq "OR" ) {
            my $src1 = $op[1];
            my $src2 = $op[2];
            my $a = evaluate($ops, $cache, $src1);
            my $b = evaluate($ops, $cache, $src2);
            $r = $a | $b;
        } elsif ( $operator eq "LSHIFT" ) {
            my $src = $op[1];
            my $amt = $op[2];
            my $a = evaluate($ops, $cache, $src);
            $r = $a << $amt;
        } elsif ( $operator eq "RSHIFT" ) {
            my $src = $op[1];
            my $amt = $op[2];
            my $a = evaluate($ops, $cache, $src);
            $r = $a >> $amt;
        }
        my $masked = $r & 0xffff;
        $cache->{$expr} = $masked;
        return $masked;
    }
}

sub main {
    if ( @ARGV < 1 ) {
        usage();
    }
    my $filename = $ARGV[0];
    my $result = process($filename);
    print "result = $result\n";
}

main();
